packages/xsnap/src/avaAssertXS.js

```diff
@@ -0,0 +1,351 @@
+        throw new Error(`missing property ${String(prop)}`);
+      }
+    }
+
+    return true;
+  }
+  const detail = JSON.stringify({
+    actual: { type: typeof x, value: x },
+    expected: { type: typeof y, value: y },
+  });
+  throw new Error(detail);
+}
+
+/**
+ * Test status reporting inspired by Test Anything Protocol (TAP)
+ *
+ * ref https://testanything.org/tap-specification.html
+ *
+ * @param {(msg: TapMessage) => void} send
+ *
+ * @typedef { ReturnType<typeof tapFormat> } TapFormat
+ * @typedef {import('./avaXS').TapMessage} TapMessage
+ */
+function tapFormat(send) {
+  return freeze({
+    /** @type {(qty: number) => void} */
+    plan(qty) {
+      send({ plan: qty });
+    },
+    /** @type {(n: number, t?: string) => void} */
+    ok(testNum, txt) {
+      send({ status: 'ok', id: testNum, message: txt });
+    },
+    /** @type {(n: number, t: string) => void} */
+    skip(testNum, txt) {
+      send({ status: 'SKIP', id: testNum, message: txt });
+    },
+    /** @type {(n: number, t?: string) => void} */
+    notOk(testNum, txt) {
+      send({ status: 'not ok', id: testNum, message: txt });
+    },
+    /** @type {(t: string, label?: string) => void} */
+    diagnostic(txt, label) {
+      send({ note: txt, label });
+    },
+  });
+}
+
+/** @type { Harness | null } */
+let theHarness = null; // ISSUE: ambient
+
+/**
+ * @param {(msg: TapMessage) => void} send
+ *
+ * @typedef { ReturnType<typeof createHarness>} Harness
+ */
+function createHarness(send) {
+  let testNum = 0;
+  let passCount = 0;
+  /** @type {((ot: { context: Object }) => Promise<void>)[]} */
+  let beforeHooks = [];
+  /** @type {(() => Promise<void>)[]} */
+  let suitesToRun = [];
+  const context = {};
+
+  /**
+   * @returns { Summary }
+   * @typedef {import('./avaXS').Summary} Summary
+   */
+  function summary() {
+    return {
+      pass: passCount,
+      fail: testNum - passCount,
+      total: testNum,
+    };
+  }
+
+  const it = freeze({
+    send,
+    get context() {
+      return context;
+    },
+    /** @type {(l: string, f: () => Promise<void>) => void } */
+    before(_label, hook) {
+      beforeHooks.push(hook);
+    },
+    /** @type { (ok: boolean) => number } */
+    finish(ok) {
+      testNum += 1;
+      if (ok) {
+        passCount += 1;
+      }
+      return testNum;
+    },
+    /** @type { (f: () => Promise<void>) => Promise<void> } */
+    async defer(thunk) {
+      suitesToRun.push(thunk);
+    },
+    summary,
+    async result() {
+      for await (const hook of beforeHooks) {
+        await hook({ context });
+      }
+      beforeHooks = [];
+      for await (const suite of suitesToRun) {
+        await suite();
+      }
+      suitesToRun = [];
+
+      return summary();
+    },
+  });
+
+  if (!theHarness) {
+    theHarness = it;
+  }
+  return it;
+}
+
+/**
+ * @param {*} exc
+ * @param {Expectation} expectation
+ *
```

Comment:
```suggestion
 * @returns {boolean}
```

since you only check the result for truthiness, should the result actually be a `boolean` instead? See below.

Reply:
tsc infers a `boolean` return type. I guess it doesn't hurt to be explicit.

My usual practice is to only add annotations until tsc stops falling back to `any`. I'm somewhat spoiled by flow (and ocaml) where much more of the types are inferred.